rope/base/oi/type_hinting/evaluate.py

```diff
@@ -0,0 +1,393 @@
+        def led(self, left, parser):
+            self.first = left
+            self.second = parser.expression(bp)
+            return self
+
+    @multi
+    def infix_r(self, name, bp):
+        symbol = self.symbol(name, bp)
+
+        @method(symbol)
+        def led(self, left, parser):
+            self.first = left
+            self.second = parser.expression(bp - 0.1)
+            return self
+
+    def ternary(self, name, name2, bp):
+        symbol = self.symbol(name, bp)
+        symbol2 = self.symbol(name2)
+
+        @method(symbol)
+        def led(self, left, parser):
+            self.first = left
+            self.second = parser.expression(symbol2.lbp)
+            parser.advance(symbol2.name)
+            self.third = parser.expression(symbol2.lbp + 0.1)
+            return self
+
+    @multi
+    def prefix(self, name, bp):
+        symbol = self.symbol(name, bp)
+
+        @method(symbol)
+        def nud(self, parser):
+            self.first = parser.expression(bp)
+            return self
+
+    @multi
+    def postfix(self, name, bp):
+        symbol = self.symbol(name, bp)
+
+        @method(symbol)
+        def led(self, left, parser):
+            self.first = left
+            return self
+
+    multi = staticmethod(multi)
+
+symbol_table = SymbolTable()
+
+
+class Lexer(object):
+
+    _token_pattern = re.compile(r"""
+        \s*
+        (?:
+              (
+                    [,()\[\]|]
+                  | ->
+                  | (?<=\s)(?:or)\b
+              )  # operator
+            | ([a-zA-Z](?:\w|\.)*)  # name
+        )
+        """, re.U | re.I | re.S | re.X)
+
+    def __init__(self, symbol_table):
+        self.symbol_table = symbol_table
+
+    def tokenize(self, program):
+        for name, value in self._tokenize_expr(program):
+            symbol = symbol_table.get(value.lower())
+            if symbol:
+                s = symbol()
+            elif name == "(name)":
+                symbol = symbol_table[name]
+                s = symbol()
+                s.value = value
+            else:
+                raise SyntaxError("Unknown operator ({0}). Possible operators are {1!r}".format(
+                    value, list(self.symbol_table)
+                ))
+
+            yield s
+
+    def _tokenize_expr(self, program):
+        if isinstance(program, bytes):
+            program = program.decode('utf-8')
+        # import pprint; pprint.pprint(self._token_pattern.findall(program))
+        for operator, name in self._token_pattern.findall(program):
+            if operator:
+                yield '(operator)', operator
+            elif name:
+                yield '(name)', name
+            else:
+                raise SyntaxError
+        yield '(end)', '(end)'
+
+
+class Parser(object):
+
+    token = None
+    next = None
+
+    def __init__(self, lexer):
+        self.lexer = lexer
+
+    def parse(self, program):
+        generator = self.lexer.tokenize(program)
+        try:
```

Comment:
I think no, I can't. First of all, it has return value, but constructor hasn't. Program is argument of state, but lexer is dependency. I create parser by factory method, and factory aware only about dependencies.